### PR TITLE
sql,externalconn: add owner_id column to system.external_connections

### PR DESCRIPTION
**Pull request by andyyang890**
This patch adds a new `owner_id` column to the `system.external_connections`
table, which corresponds to the existing `owner` column, in order to bring us
closer to the eventual goal of allowing renaming of users. Migrations are
also added to alter and backfill the table in older clusters.

Part of #87079
Part of #92342

Release note: None

### CODE REVIEW

**Review by stevendanna — COMMENTED** [review]
Thanks! It might be good to put why we are adding this field in the PR/commit message somewhere.

We also need to handle external_connections tables in backups that don't the owner_id column.  Check out 1bc6c6d0bf54 for an example.

**Comment by andyyang890** [comment]
> It might be good to put why we are adding this field in the PR/commit message somewhere.

Done.

> We also need to handle external_connections tables in backups that don't the owner_id column. 

Yep, I'm planning on doing that in a followup PR where I set the column to be NOT NULL.

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97877)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @andyyang890, @dt, and @stevendanna)

___
*[`pkg/cloud/externalconn/record.go` line 355 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NPrx0MEExzEF9QHS43N:-NPrx0MEExzEF9QHS43O:bwqz2da) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1639fffe358696575fa0c382156ed01e971b605/pkg/cloud/externalconn/record.go#L355)):*
> ```Go
> 		case `owner`:
> 			arg = tree.NewDString(e.rec.Owner.Normalized())
> 		case `owner_id`:
> ```

it looks like this change will cause the owner_id column to always be stored. but we need to gate it on the clusterversion you just added. 

and it would be good to have a test that shows it working fine in a mixed version cluster. let's add another mixed-version logic test using the new framework (which is skipped right now, but you should be able to run locally)
___
*[`pkg/sql/create_external_connection.go` line 141 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NPrvL1O7ZAp1EqCvYSs:-NPrvL1O7ZAp1EqCvYSt:b-or4gpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1639fffe358696575fa0c382156ed01e971b605/pkg/sql/create_external_connection.go#L141)):*
> ```Go
> 	row, err := txn.QueryRowEx(params.ctx, `get-user-id`, txn.KV(),
> 		sessiondata.NodeUserSessionDataOverride,
> 		`SELECT user_id FROM system.users WHERE username = $1`,
> ```

in the future, we should add another function similar to `p.User()` that returns the user ID instead. that way we can cache the user ID in one place, and avoid having to do this lookup query in many places. could you file a github issue for that, and make a comment here that references it?

**Review by andyyang890 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97877)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @rafiss, and @stevendanna)

___
*[`pkg/sql/create_external_connection.go` line 141 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NPrvL1O7ZAp1EqCvYSs:-NPxkJQS1Yc4efAy9DFY:b-mrp7e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1639fffe358696575fa0c382156ed01e971b605/pkg/sql/create_external_connection.go#L141)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

in the future, we should add another function similar to `p.User()` that returns the user ID instead. that way we can cache the user ID in one place, and avoid having to do this lookup query in many places. could you file a github issue for that, and make a comment here that references it?
</blockquote></details>

Done. See https://github.com/cockroachdb/cockroach/issues/98170

**Review by andyyang890 — COMMENTED** [review]
FYI I rebased this onto https://github.com/cockroachdb/cockroach/pull/98163 so the first commit is from that.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-:-NPyEgAO-uUilWoHOCZi:b-ll2sg4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @rafiss, and @stevendanna)

___
*[`pkg/cloud/externalconn/record.go` line 355 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NPrx0MEExzEF9QHS43N:-NPyDvQJ43oxvAk1OcIU:b5evh39) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1639fffe358696575fa0c382156ed01e971b605/pkg/cloud/externalconn/record.go#L355)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

it looks like this change will cause the owner_id column to always be stored. but we need to gate it on the clusterversion you just added. 

and it would be good to have a test that shows it working fine in a mixed version cluster. let's add another mixed-version logic test using the new framework (which is skipped right now, but you should be able to run locally)
</blockquote></details>

I added some gating by threading through a map of excluded columns from the relevant planner function since I don't think this function has access to the version handle, but I'm not sure if this is the best approach. Let me know if you have any other ideas!
___
*[`pkg/upgrade/upgrades/external_connections_table_user_id_migration_test.go` line 20 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NPyF2MnEsN0sM9mospT:-NPyF2MnEsN0sM9mospU:b-ju2vms) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0a04d572381842de9906964cc6b67cfd8359703/pkg/upgrade/upgrades/external_connections_table_user_id_migration_test.go#L20)):*
> ```Go
> 
> 	"github.com/cockroachdb/cockroach/pkg/base"
> 	_ "github.com/cockroachdb/cockroach/pkg/cloud/userfile"
> ```

I had to add this import in order to trigger the `init()` function to make the `CREATE EXTERNAL CONNECTION` statement below work.

**Review by rafiss — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97877)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @andyyang890, @dt, and @stevendanna)

___
*[`pkg/cloud/externalconn/record.go` line 355 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NPrx0MEExzEF9QHS43N:-NQ6wHVdEm7re0o90pxF:b-8hdhch) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1639fffe358696575fa0c382156ed01e971b605/pkg/cloud/externalconn/record.go#L355)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

I added some gating by threading through a map of excluded columns from the relevant planner function since I don't think this function has access to the version handle, but I'm not sure if this is the best approach. Let me know if you have any other ideas!
</blockquote></details>

this looks like it works, but i'd be curious if @stevendanna or @adityamaru have opinions on how to version gate this, since they own this area of the code.
___
*[`pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade` line 3 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NQ6w4te7vDDv0zO8GUh:-NQ6w4te7vDDv0zO8GUi:b-v8i2tp) ([raw file](https://github.com/cockroachdb/cockroach/blob/52a82cdae55a172eeb0e5ce11b7ec5ccf5cd363f/pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade#L3)):*
> ```
> # LogicTest: cockroach-go-testserver-22.2-master
> 
> # Verify that all nodes are running 22.2 binaries.
> ```

nit: i think we can remove these checks for 22.2 binaries. they were there in the first mixed version test more as a sanity check / proof-of-concept
___
*[`pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade` line 52 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NQ6wEAy9RewfqQUY3KR:-NQ6wEAy9RewfqQUY3KS:b-z3p5ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/52a82cdae55a172eeb0e5ce11b7ec5ccf5cd363f/pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade#L52)):*
> ```
> upgrade 2
> 
> # Verify that all nodes are now running 23.1 binaries.
> ```

ditto above

**Review by adityamaru — APPROVED** [review]
Thanks for doing this!

File: pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+true
+
+query B nodeidx=2
+SELECT crdb_internal.node_executable_version() SIMILAR TO '1000022.2-%'
+----
+true
+
+# Wait for migrations to run and verify that owner_id column is now present.
+
+sleep 10s
```
[the remark this answers is not in the document]
Comment (andyyang890) [reply pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade]: I guess not, I've removed it!

**Review by andyyang890 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97877)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @dt, @rafiss, and @stevendanna)

___
*[`pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade` line 3 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NQ6w4te7vDDv0zO8GUh:-NQ94kSP6mRpAFpJljUX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52a82cdae55a172eeb0e5ce11b7ec5ccf5cd363f/pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade#L3)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: i think we can remove these checks for 22.2 binaries. they were there in the first mixed version test more as a sanity check / proof-of-concept
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade` line 52 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/97877#-NQ6wEAy9RewfqQUY3KR:-NQ94jWBDN2omJnfrZIw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52a82cdae55a172eeb0e5ce11b7ec5ccf5cd363f/pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade#L52)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

ditto above
</blockquote></details>

Done.

**Comment by andyyang890** [comment]
TFTRs!

bors r=rafiss,adityamaru

**Comment by andyyang890** [comment]
bors r=rafiss,adityamaru

File: pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade (hunk 1 above)
[the remark this answers is not in the document]
Comment (andyyang890) [reply pkg/sql/logictest/testdata/logic_test/testserver_22.2_23.1_external_connections_user_id_upgrade]: Actually, it looks like the test is flaky on CI without the additional sleep so I'm adding it back in https://github.com/cockroachdb/cockroach/pull/98401.

**Pull request merged**